internal/uvm/clone.go

```diff
@@ -0,0 +1,95 @@
+package uvm
+
+import (
+	"context"
+
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+	"github.com/pkg/errors"
+)
+
+const hcsSaveOptions = `{"SaveType": "AsTemplate"}`
+
+// Cloneable is a generic interface for cloning a specific resource. Not all resources can
+// be cloned and so all resources might not implement this interface. This interface is
+// mainly used during late cloning process to clone the resources associated with the UVM
+// and the container. For some resources (like scratch VHDs of the UVM & container)
+// cloning means actually creating a copy of that resource while for some resources it
+// simply means adding that resource to the cloned VM without copying (like VSMB shares).
+// The Clone function of that resource will deal with these details.
+type Cloneable interface {
+	// Clone function creates a clone of the resource on the UVM `vm` and returns a
+	// pointer to the struct that represents the cloned resource.
+	// `cd` parameter can be used to pass any other data that is required during the
+	// cloning process of that resource (for example, when cloning SCSI Mounts we
+	// might need scratchFolder).
+	// Clone function should be called on a valid struct (Mostly on the struct which
+	// is deserialized, and so Clone function should only depend on the fields that are
+	// exported in the struct).
+	// The implementation of the clone function should avoid reading any data from the
+	// `vm` struct, it can add new fields to the vm struct but since the vm struct
+	// isn't fully ready at this point it shouldn't be used to read any data.
+	Clone(ctx context.Context, vm *UtilityVM, cd *cloneData) (interface{}, error)
+}
+
+// cloneData contains all the information that might be required during cloning process of
+// a resource.
+type cloneData struct {
+	// doc spec for the clone
+	doc *hcsschema.ComputeSystem
+	// scratchFolder of the clone
+	scratchFolder string
+	// UVMID of the clone
+	uvmID string
+}
+
+// UVMTemplateConfig is just a wrapper struct that keeps together all the resources that
+// need to be saved to create a template.
+type UVMTemplateConfig struct {
+	// ID of the template vm
+	UVMID string
+	// Array of all resources that will be required while making a clone from this template
+	Resources []Cloneable
+}
+
+// Captures all the information that is necessary to properly save this UVM as a template
+// and create clones from this template later. The struct returned by this method must be
+// later on made available while creating a clone from this template.
+func (uvm *UtilityVM) GenerateTemplateConfig() *UVMTemplateConfig {
+	// Add all the SCSI Mounts and VSMB shares into the list of clones
+	templateConfig := &UVMTemplateConfig{
+		UVMID: uvm.ID(),
+	}
+
+	for _, vsmbShare := range uvm.vsmbDirShares {
+		templateConfig.Resources = append(templateConfig.Resources, vsmbShare)
```

Comment:
I don't think we should encode the resources directly, as that seems like (especially if we add more complex resource types) it could lead to issues where the rehydrated resource tries to rely on state that is no longer valid.

Instead, could we define a somewhat more scoped set of things to write out, and focus on just including enough information that it allows us to create a replacement resource in the cloned UVM?

Reply:
I went with the idea that only the exported fields are saved and for SCSI and VSMB resource types the exported fields don't include any state. However I agree that if we modify these resources or add new resources in the future then we might end up saving some state of a resource.
I think we can add a separate saved state type for each resource and store that in the registry. It shouldn't be too difficult to override the encoding/decoding functions for each resource to generate this new data type. Does that make sense?